### DISCUSSION TITLE
Stringnames without double quotes

### DISCUSSION THREAD

**Discussion by Agecaf** [godotengine/godot-proposals#10854] [Scripting]
If I understand correctly, StringNames have a literal form in GDScript, so writing `&"ui_accept"` generates the StringName for accepting ui. What I'm thinking is that it might be nice to be able to also write `&ui_accept` to generate the same StringName. 

StringNames are often used for strings that do not have whitespace inside them, like names of methods in `Object.call` or action names in `Input.is_action_just_pressed`.

Currently writing `&ui_accept` syntax highlights it the same color as a proper StringName like `&"ui_accept"`, even if `&ui_accept` is a syntax error. So from a parsing perspective it should be a relatively simple addition. This also means that if you forget you need double quotes on StringNames, you run into a syntax error

There may be confusion with the binary `&` operator, but usually string names, like identifiers, do not start with a digit, and the syntax highlighter seems fairly good at detecting when an `&` is being used as a binary operator or as the start of a StringName. If string names starting with digits or with spaces are required, users would have no issue using the previous notation as `&"42"` or `&"Hello World!"`.

This behavior would be similar to the not-literal `$` for accessing nodes, which can be currently used as `$ChildNode` or `$"ChildNode"` interchangeably.

**Comment by dalexeev** [comment]
See:

* #4721 
* #8079

**Comment by Mickeon** [comment]
Actually, see dalexeev's own comment on the matter in https://github.com/godotengine/godot-proposals/issues/4721#issuecomment-1759307302

**Comment by Agecaf** [comment]
Thanks for the clarification! I wasn't able to find these discussions in my first search. I thought the reason that the StringName operator was chosen to be like `&"this"` (as opposed to something like `s"this"` similar to `r"raw strings"`) was to make it similar to pointers/references in C, since StringNames are often used as references to method names and input action names.

StringNames if I understand correctly are the equivalent of [keywords in Clojure](https://cljs.github.io/api/syntax/keyword) or [symbols in Ruby](https://ruby-doc.org/core-2.5.1/Symbol.html), both of which use a notation like `:this` (with Ruby also having a notation like `:"this"`), they don't seem to require the double quotes in those cases from a usability perspective. 

I think there's a usability for users familiar with StringNames vs confusion for users not-yet-familiar with StringNames balance to be found, but this could be addressed from a documentation perspective. Nodepaths like `$this` made me initially confused that GDScript was going to be like Perl, but that was quickly dispelled once I learnt what Nodepaths were and how the notation was super useful.

**Reply by dalexeev** [reply]
Another important difference between the `$Node` construct and the `&"StringName"` and `^"NodePath"` literals is that if you make a typo in the node name, you'll get a runtime error immediately, whereas strings can contain any content and the error may occur much later.

If you see `&my_method`, you might incorrectly think that this syntax provides some compile-time or at least runtime guarantees that `my_method` actually exists (like first-class callables and signals do). See also `nameof` proposals (#837, #10117). In my opinion, the quotes more explicitly tell you "hey, I'm just a string (albeit a special one) and can contain anything".
